README.md
# esp32_python_eq3
Bluetooth bridge between eqiva EQ3 thermostat and WLAN network, using ESP32, micropython and mqtt.

 - INSTRUCTION https://github.com/yunnanpl/esp32_python_eq3/blob/master/INSTRUCTION.md
 - CHANGELOG https://github.com/yunnanpl/esp32_python_eq3/blob/master/CHANGELOG.md

 - https://www.eq-3.de/produkte/eqiva/detail/bluetooth-smart-heizkoerperthermostat.html

# Idea
I used a lot of other solutions, but it seemed to be not stable for me, or hard to extend (for me :)).
In order to use the esp32 for EQ3 and Mijia temperature sensors (or even some BME280-like ones) I needed to get something done on my own.

So I decided to rewrite it in python (micropython) on esp32.

This is the result. It can mimic and substitute the esp32_mqtt_eq3 solution from softypit, as it accepts the same mqtt input and returns similar mqtt json output.

# Status

It works quite stable. More stable for me than esp32_mqtt_eq3 solution.
I added also work list (so multpiple work can be added, and it will be done one after another).
This solution reads data from Mijia sensors. Plan is to add BME280 and other "local" sensors.
It also does BLE scans, and can report bluetooth devices in the area (for presence testing etc.).

Communication is done by MQTT, and the syntax is near to identical to esp32_mqtt_eq3 so it can be substituted.
The syntax is still simpler than esp32_mqtt_eq3.

Serial or WebREPL can be used. THere are some useful functions on the device.

## Tests
The aim is stability.
Test system is 4 ESP32 controlling 9 EQ3 (around 2 EQ3 for each ESP32 and reading out 2 thermometers).
Anything can happen: Battery on EQ3 runs out. BLE packets get corrupted, etc... WLAN breaks, MQTT server dies, bad work is sent... DOS happens...
I tried to pick up all imaginable errors. I do up to 700 queries daily (every 2-3 minutes), and send 10-50 commands.
At best, the device should never loose work, even if it is forced sometimes to reboot.

Tests with newest version... 
 - 48h, 1400 queries, 30 commands for each of EQ3, configuration as written above -> no issues
 - further testing

It seems that RSSI above -90 leads to issues with connection. It will connect, but after a few tries.
Below -90 it works without any issues.

## Features (and anti-features)
 - NEW basic webpage added
 - NO AP configuration (only config file)
 - robust WIFI, survives disconnections, reconnects, etc.
 - robust BLE, survives bad addresses, bad messages, disconnections, etc. Has automatic connections cleaning.
 - robust MQTT, survives disconnections, automatically resubscribes, etc.
   - more robust mqtt, as there were issues that mqtt-robust lib, have not noticed that it was not subscribed
 - has "unique" worklist
   - if the same message will be sent 5 times, it will be processed once
   - if there is a not-yet-processed work in the list for setting temp, then another work for the same device overwrites previous command
 - is persistent, if the device is known, but does not respond in 10-20 sec slot, then work is put at the end, and new work is done 
   
## Plan
 - add basic AP for configuration
 - add other temperature sensors like Mijia, BME280, etc.
 - add PID regulator for automatic temperature regulation
   - in case of real temperature input
 - add other basic info, like battery, mode, etc.
 - add offset setting, both in eq3 and "external" through esp32
 - think about adding support for the new firmware version of eq3 (later :D)
   - maybe not so tragic at all, it has to do with the new pairing
   - https://github.com/rytilahti/python-eq3bt/issues/41
   - and micropython ble seems to support pairing already
   - https://github.com/micropython/micropython/pull/6651
 - maybe - add temperature history and eq3 setting history graph

# References

 - mqtt - most robust library chosen
   - https://github.com/fizista/micropython-umqtt.simple2 with https://github.com/fizista/micropython-umqtt.robust2
   - not clear if necessary

# Other interesting liks
 - https://github.com/softypit/esp32_mqtt_eq3
 - https://github.com/rytilahti/python-eq3bt
 - https://github.com/bimbar/bluepy_devices


CHANGELOG.md
# Features
   + controlling and querying of EQ3 devices
     - temperature can be set and queried
     - modes are not yet available (battery also not)
   + readout of Mijia thermometers
   + communication as MQTT (partial compatibilty to esp32_mqtt_eq3 from softypit)
   + basic webpage interface
     + running on async instead of sockets now
   + added basic OTA !!!
     + files can be uploaded and removed from ESP through the page
     + works for any text file, does not work for binary files yet...
   + removed necessity of webrepl to save memory
     + webrepl can be switched on and off through the webpage
   + improved web/sessions - more responsive
   + improved links (upload and reset requires confirmation)
   + added more clear descriptions
   - precompiled the external libraries to save memory and space (simple2.py)
     + precompilation was removed, as for every micropython version a new precompiled file might be needed

# Changes
 - planned
   - BLE detections with lower signal than -92 are ignored
 - 7a5f011
   - EQ3 now sends valve open value and battery status (also with home assistant autodiscovery
   - web OTA and some other pages, postpone BLE actions to improve responsibility of the webpage
   - precompiled MPY files are dropped, due to issues with compatibility with different micropython versions
   - all files (also config and simple2) and outputs (page) are now latin-1 encoded (not utf-8)
     - this seems to lower and stabilize the memory use, at the cost of using special characters
   - presence for BLE beacons (or some smart watches, like Mi Band) detection is added, but works unstable
   - temperature range selection is expanded to full EQ3 range (5 to 29.5)
     - it does not always work at the start, as new autodiscovery messages need to be sent
   - BUG: wrong byte for valve open was extracted
 - aefe81f
   - significant code rewrite (cleaning pending)
   - automatic time NTP querying
   - automatic MQTT autodiscovery for Home Assistant
     - for thermostats (climate entity)
     - for thermometers (sensor entity)
   - work list for each device separately
   - use of multiple paralel BLE connections, for speed and stability
   - whitelist for EQ3, thermometers-sensors
   - automatic querying of EQ3 (and thermometers, if available)
   - async webpage for more stability
   - watchdog for rebooting just in case
   - BUG: when the contacted device was not available, then the system stuck in a loop (as it retried the connection all the time)
     - now there are 3 retries in a row, and then it moves to the next work
   - BUG: there was a 1 second slot, in which the old work was done, and new work could be added, and quickly deleted by the previous deletion request
     - this cannot happen now, as there is a work list
   - BUG: whitelisted devices could not be removed from GUI
     - now possible
   - BUG Theoretical: If the connections will be so fast, and amount of the EQ3s larger than 6-7, then all BLE connections will be saturated, and it is not known what will happen
     - now, if 4 simultanuous connections are opened, then ESP32 is waiting for done connections to be closed, before opening new
   - BUG: Thermometer readout timer was reset so often, that the real readout never happened (as it was always "just done")
     - fixed
   - BUG: NTP and Autodiscovery were done every few minutes, as the work timer was not set correctly (NTP every 24 hours, Autodiscovery every 6 hours)
     - fixed
   - ISSUE: worklist could consist of multiple equal commands, which makes no sense
     - now if equal command exists in the list, it is dropped, except this, no other commands are dropped
   - ISSUE: flooding web connections could comsume all the RAM and lead to reboot
     - now it should not happen, as the device will respond "flood" or not respond at all if free RAM amount is critically low, but it will try to stay alive
   - ISSUE: it is more an issue, as scanning could be interrupted by any BLE connection request, and scanning did not start when BLE connection was open
     - this led to an issue, where the scanning was very short, or did not took place for a few minutes in a row
     - now, the scanning reserves 1/2 of the full requested scan time, so in worst case it takes half of time, or in best complete time
   - ISSUE: it was hard to controll the ESP32/EQ3 externally, as the MQTT addresses were not known
     - now in the Info page there is a list of the devices, current work/status and a list of MQTT addresses to control the system externally (not through Home Assistant)
 - previous
   - whatever


code/boot.py
# -*- coding: latin-1 -*-
# -### boot.py

# Done by Dr.JJ
# https://github.com/yunnanpl/esp32_python_eq3

#-###
#-###
# -### imports, also CONFIG and settings
#
import uasyncio as asyncio
import _thread
import network
import ntptime
import time
import ubluetooth
import gc
import simple2 as umqtt
#import robust2 as umqtt
#from collections import OrderedDict
import os
from secret_cfg import *
# speedup/slow down for energy saving :)
import machine
machine.freq(CONFIG['freq'])
#
# import re
# import robust2 as umqtt
# import socket
# not necessary
# from micropython import const
#
# other modules
# from machine import Pin, DAC, PWM, ADC, SoftI2C
# from machine import Pin
# import umqtt
# import math

gc.enable()

#-###
#-### CONFIG compatibility
# moving variable to list of permanent variables, in order to query ntp sometimes
# and not to change CONFIG
try:
    CONFIG2['mqtt_thermo_src'] = CONFIG2['mqtt_temp_src']
except:
    pass

try:
    CONFIG2['ntp_host'] = CONFIG['ntp_host']
except:
    pass

#-###
#-###
# -### activate ble
ble = ubluetooth.BLE()
if ble.active() == False:
    ble.active(True)

#-###
#-###
# -### create timers
#timer_schedule = machine.Timer(0)
timer_work = machine.Timer(0)
timer_check = machine.Timer(1)

#-###
#-###
# -### conenct to network
try:
    station = network.WLAN(network.STA_IF)
    station.active(True)
    # station.connect( CONFIG['wifi_name'], binascii.a2b_base64( CONFIG['wifi_pass'] ) )
    # CONFIG['wifi_name'] = "asdasd"
    station.connect(CONFIG['wifi_name'], "".join([chr(x) for x in CONFIG['wifi_pass']]))
except:
    pass

#-###
#-###
# -### waiting for connection
while station.isconnected() == False:
    # print('LOG waiting for connection')
    time.sleep(0.5)  # sleep or pass
    # if not conencted in 15 sec, then restart
    if time.ticks_ms() > 15000:
        print('- no WIFI -> reset')
        machine.reset()

#-###
#-###
# -### getting ntp time
try:
    ntptime.host = CONFIG2['ntp_host']
    ntptime.settime()
except:
    pass
# print('LOG NTP time set')

try:
    # if webrepl_cfg file exists then start webrepl
    # if not, then pass
    os.stat('webrepl_cfg.py')
    import webrepl
    webrepl.start()
except:
    pass

try:
    mqtth = umqtt.MQTTClient(CONFIG2['mqtt_usr'], CONFIG2['mqtt_srv'], user=CONFIG2['mqtt_usr'], password=CONFIG2['mqtt_pass'], port=1883)
    mqtth.keepalive = 180
    while mqtth.connect() == 1:
        time.sleep(0.5)
        # if not conencted in 20 sec, then restart
        if time.ticks_ms() > 20000:
            print('- no MQTT -> reset')
            machine.reset()
    mqtth.ping()
    mqtth.set_callback(fmqtt_irq)
    # keepalive describes the connection lifetime, 0 means not defined
    #mqtth.keepalive = 1
except:
    pass

#-###
#-### some other settings

# ble.config(rxbuf=256)
#ble.config(rxbuf=512)
#ble.config(mtu=515)
# ble.config(gap_name=b'ESP32_5')
#ble.config( gap_name='ESP32_'+str(ble.config('mac')[1][5]))
device_name = 'ESP32_' + str(station.ifconfig()[0].split('.')[3])
ble.config(gap_name=device_name)

# -### clean up the memory and stuff
CONFIG = ''
del CONFIG
# a lot of garbage expected :D

print('= ip', str( station.ifconfig()[0] ) )
print('= time', ntptime.time() )
print('= mqtt ping', str(time.ticks_ms() - mqtth.last_cpacket) )
print('+ booted')

gc.collect()
gc.threshold(40000) # ???
# -### BOOTED
# -### end


code/simple2.py
# -*- coding: latin-1 -*-
import usocket as socket
import uselect
from utime import ticks_add,ticks_ms,ticks_diff
class MQTTException(Exception):0
def pid_gen(pid=0):
	A=pid
	while True:A=A+1 if A<65535 else 1;yield A
class MQTTClient:
	def __init__(A,client_id,server,port=0,user=None,password=None,keepalive=0,ssl=False,ssl_params=None,socket_timeout=5,message_timeout=10):
		C=ssl_params;B=port
		if B==0:B=8883 if ssl else 1883
		A.client_id=client_id;A.sock=None;A.poller=None;A.server=server;A.port=B;A.ssl=ssl;A.ssl_params=C if C else{};A.newpid=pid_gen()
		if not getattr(A,'cb',None):A.cb=None
		if not getattr(A,'cbstat',None):A.cbstat=lambda p,s:None
		A.user=user;A.pswd=password;A.keepalive=keepalive;A.lw_topic=None;A.lw_msg=None;A.lw_qos=0;A.lw_retain=False;A.rcv_pids={};A.last_ping=ticks_ms();A.last_cpacket=ticks_ms();A.socket_timeout=socket_timeout;A.message_timeout=message_timeout
	def _read(A,n):
		try:
			B=b''
			for C in range(n):A._sock_timeout(A.poller_r,A.socket_timeout);B+=A.sock.read(1)
		except AttributeError:raise MQTTException(8)
		if B==b'':raise MQTTException(1)
		if len(B)!=n:raise MQTTException(2)
		return B
	def _write(A,bytes_wr,length=-1):
		D=bytes_wr;B=length
		try:A._sock_timeout(A.poller_w,A.socket_timeout);C=A.sock.write(D,B)
		except AttributeError:raise MQTTException(8)
		if B<0:
			if C!=len(D):raise MQTTException(3)
		elif C!=B:raise MQTTException(3)
		return C
	def _send_str(A,s):assert len(s)<65536;A._write(len(s).to_bytes(2,'big'));A._write(s)
	def _recv_len(D):
		A=0;B=0
		while 1:
			C=D._read(1)[0];A|=(C&127)<<B
			if not C&128:return A
			B+=7
	def _varlen_encode(C,value,buf,offset=0):
		B=offset;A=value;assert A<268435456
		while A>127:buf[B]=A&127|128;A>>=7;B+=1
		buf[B]=A;return B+1
	def _sock_timeout(B,poller,socket_timeout):
		A=socket_timeout
		if B.sock:
			C=poller.poll(-1 if A is None else int(A*1000))
			if not C:raise MQTTException(30)
		else:raise MQTTException(28)
	def set_callback(A,f):A.cb=f
	def set_callback_status(A,f):A.cbstat=f
	def set_last_will(A,topic,msg,retain=False,qos=0):B=topic;assert 0<=qos<=2;assert B;A.lw_topic=B;A.lw_msg=msg;A.lw_qos=qos;A.lw_retain=retain
	def connect(A,clean_session=True):
		E=clean_session;A.sock=socket.socket();A.poller_r=uselect.poll();A.poller_r.register(A.sock,uselect.POLLIN);A.poller_w=uselect.poll();A.poller_w.register(A.sock,uselect.POLLOUT);G=socket.getaddrinfo(A.server,A.port)[0][-1];A.sock.connect(G)
		if A.ssl:import ussl;A.sock=ussl.wrap_socket(A.sock,**A.ssl_params)
		F=bytearray(b'\x10\x00\x00\x00\x00\x00');B=bytearray(b'\x00\x04MQTT\x04\x00\x00\x00');D=10+2+len(A.client_id);B[7]=bool(E)<<1
		if bool(E):A.rcv_pids.clear()
		if A.user is not None:
			D+=2+len(A.user);B[7]|=1<<7
			if A.pswd is not None:D+=2+len(A.pswd);B[7]|=1<<6
		if A.keepalive:assert A.keepalive<65536;B[8]|=A.keepalive>>8;B[9]|=A.keepalive&255
		if A.lw_topic:D+=2+len(A.lw_topic)+2+len(A.lw_msg);B[7]|=4|(A.lw_qos&1)<<3|(A.lw_qos&2)<<3;B[7]|=A.lw_retain<<5
		H=A._varlen_encode(D,F,1);A._write(F,H);A._write(B);A._send_str(A.client_id)
		if A.lw_topic:A._send_str(A.lw_topic);A._send_str(A.lw_msg)
		if A.user is not None:
			A._send_str(A.user)
			if A.pswd is not None:A._send_str(A.pswd)
		C=A._read(4)
		if not(C[0]==32 and C[1]==2):raise MQTTException(29)
		if C[3]!=0:
			if 1<=C[3]<=5:raise MQTTException(20+C[3])
			else:raise MQTTException(20,C[3])
		A.last_cpacket=ticks_ms();return C[2]&1
	def disconnect(A):A._write(b'\xe0\x00');A.poller_r.unregister(A.sock);A.poller_w.unregister(A.sock);A.sock.close();A.sock=None;A.poller=None
	def ping(A):A._write(b'\xc0\x00');A.last_ping=ticks_ms()
	def publish(A,topic,msg,retain=False,qos=0,dup=False):
		E=topic;B=qos;assert B in(0,1);C=bytearray(b'0\x00\x00\x00\x00');C[0]|=B<<1|retain|int(dup)<<3;F=2+len(E)+len(msg)
		if B>0:F+=2
		G=A._varlen_encode(F,C,1);A._write(C,G);A._send_str(E)
		if B>0:D=next(A.newpid);A._write(D.to_bytes(2,'big'))
		A._write(msg)
		if B>0:A.rcv_pids[D]=ticks_add(ticks_ms(),A.message_timeout*1000);return D
	def subscribe(A,topic,qos=0):E=topic;assert qos in(0,1);assert A.cb is not None,'Subscribe callback is not set';B=bytearray(b'\x82\x00\x00\x00\x00\x00\x00');C=next(A.newpid);F=2+2+len(E)+1;D=A._varlen_encode(F,B,1);B[D:D+2]=C.to_bytes(2,'big');A._write(B,D+2);A._send_str(E);A._write(qos.to_bytes(1,'little'));A.rcv_pids[C]=ticks_add(ticks_ms(),A.message_timeout*1000);return C
	def _message_timeout(A):
		C=ticks_ms()
		for (B,D) in A.rcv_pids.items():
			if ticks_diff(D,C)<=0:A.rcv_pids.pop(B);A.cbstat(B,0)
	def check_msg(A):
		if A.sock:
			if not A.poller_r.poll(-1 if A.socket_timeout is None else 1):A._message_timeout();return None
			try:
				G=A._read(1)
				if not G:A._message_timeout();return None
			except OSError as H:
				if H.args[0]==110:A._message_timeout();return None
				else:raise H
		else:raise MQTTException(28)
		if G==b'\xd0':
			if A._read(1)[0]!=0:MQTTException(-1)
			A.last_cpacket=ticks_ms();return
		B=G[0]
		if B==64:
			D=A._read(1)
			if D!=b'\x02':raise MQTTException(-1)
			F=int.from_bytes(A._read(2),'big')
			if F in A.rcv_pids:A.last_cpacket=ticks_ms();A.rcv_pids.pop(F);A.cbstat(F,1)
			else:A.cbstat(F,2)
		if B==144:
			C=A._read(4)
			if C[0]!=3:raise MQTTException(40,C)
			if C[3]==128:raise MQTTException(44)
			if C[3]not in(0,1,2):raise MQTTException(40,C)
			E=C[2]|C[1]<<8
			if E in A.rcv_pids:A.last_cpacket=ticks_ms();A.rcv_pids.pop(E);A.cbstat(E,1)
			else:raise MQTTException(5)
		A._message_timeout()
		if B&240!=48:return B
		D=A._recv_len();I=int.from_bytes(A._read(2),'big');J=A._read(I);D-=I+2
		if B&6:E=int.from_bytes(A._read(2),'big');D-=2
		K=A._read(D)if D else b'';L=B&1;M=B&8;A.cb(J,K,bool(L),bool(M));A.last_cpacket=ticks_ms()
		if B&6==2:A._write(b'@\x02');A._write(E.to_bytes(2,'big'))
		elif B&6==4:raise NotImplementedError()
		elif B&6==6:raise MQTTException(-1)
	def wait_msg(A):B=A.socket_timeout;A.socket_timeout=None;C=A.check_msg();A.socket_timeout=B;return C

code/secret_cfg.py
# -*- coding: latin-1 -*-
#-### secret_cfg.py

#-### variables / configuration

#-###
#-###
#-### there variables are used only during boot, then, removed from memory
CONFIG = {}
CONFIG['freq'] = 240000000
#CONFIG['freq'] = 160000000
#-### pass is convoluted with [ ord(x) for x in list(pass) ]
CONFIG['wifi_pass'] = [112, 97, 115, 115]
CONFIG['wifi_name'] = 'wifiname'
CONFIG['ntp_host'] = 'ntp_server'
#
CONFIG2 = {}
CONFIG2['mqtt_eq3_in'] = '/hass/climate/climate/radin/trv'
CONFIG2['mqtt_eq3_out'] = '/hass/climate/climate/radout/status'
CONFIG2['mqtt_mijia_out'] = '/hass/climate/thermo/status'

CONFIG2['mqtt_eq3'] = 'esp/climate/'
CONFIG2['mqtt_thermo'] = 'esp/thermo/'
# source of real temperature, some other mqtt thermometer or something
CONFIG2['mqtt_thermo_src'] = ''
CONFIG2['mqtt_presence'] = 'esp/presence/'

CONFIG2['mqtt_srv'] = 'mqtt_srv'
CONFIG2['mqtt_usr'] = 'mqtt_usr'
CONFIG2['mqtt_pass'] = 'mqtt_pass'

#-### if keep loop is set to 0, then loops die
#-### if you want to kill loops/threads, set this to 0
#-### good for debugging and responsiver webrepl
CONFIG2['loop'] = 1

#-### done here

code/robust2.py
from utime import ticks_ms, ticks_diff
#from . import simple2
import simple2

class MQTTClient(simple2.MQTTClient):
    DEBUG = False

    # Information whether we store unsent messages with the flag QoS==0 in the queue.
    KEEP_QOS0 = True
    # Option, limits the possibility of only one unique message being queued.
    NO_QUEUE_DUPS = True
    # Limit the number of unsent messages in the queue.
    MSG_QUEUE_MAX = 5
    # When you reconnect, all existing subscriptions are renewed.
    RESUBSCRIBE = True

    def __init__(self, *args, **kwargs):
        """
        See documentation for `umqtt.simple2.MQTTClient.__init__()`
        """
        super().__init__(*args, **kwargs)
        self.subs = []  # List of stored subscriptions
        self.msg_to_send = []  # Queue with list of messages to send
        self.sub_to_send = []  # Queue with list of subscriptions to send
        self.msg_to_confirm = {}  # Queue with a list of messages waiting for the server to confirm of the message.
        self.sub_to_confirm = {}  # Queue with a subscription list waiting for the server to confirm of the subscription
        self.conn_issue = None  # We store here if there is a connection problem.

    def is_keepalive(self):
        """
        It checks if the connection is active. If the connection is not active at the specified time,
        saves an error message and returns False.

        :return: If the connection is not active at the specified time returns False otherwise True.
        """
        time_from__last_cpackage = ticks_diff(ticks_ms(), self.last_cpacket) // 1000
        if 0 < self.keepalive < time_from__last_cpackage:
            self.conn_issue = (simple2.MQTTException(7), 9)
            return False
        return True

    def set_callback_status(self, f):
        """
        See documentation for `umqtt.simple2.MQTTClient.set_callback_status()`
        """
        self._cbstat = f

    def cbstat(self, pid, stat):
        """
        Captured message statuses affect the queue here.

        stat == 0 - the message goes back to the message queue to be sent
        stat == 1 or 2 - the message is removed from the queue
        """
        try:
            self._cbstat(pid, stat)
        except AttributeError:
            pass

        for data, pids in self.msg_to_confirm.items():
            if pid in pids:
                if stat == 0:
                    self.msg_to_send.insert(0, data)
                pids.remove(pid)
                if not pids:
                    self.msg_to_confirm.pop(data)
                return

        for data, pids in self.sub_to_confirm.items():
            if pid in pids:
                if stat == 0:
                    self.sub_to_send.append(data)
                pids.remove(pid)
                if not pids:
                    self.sub_to_confirm.pop(data)

    def connect(self, clean_session=True):
        """
        See documentation for `umqtt.simple2.MQTTClient.connect()`.

        If clean_session==True, then the queues are cleared.

        Connection problems are captured and handled by `is_conn_issue()`
        """
        if clean_session:
            self.msg_to_send[:] = []
            self.msg_to_confirm.clear()
        try:
            out = super().connect(clean_session)
            self.conn_issue = None
            return out
        except (OSError, simple2.MQTTException) as e:
            self.conn_issue = (e, 1)

    def log(self):
        if self.DEBUG:
            if type(self.conn_issue) is tuple:
                conn_issue, issue_place = self.conn_issue
            else:
                conn_issue = self.conn_issue
                issue_place = 0
            place_str = ('?', 'connect', 'publish', 'subscribe',
                         'reconnect', 'sendqueue', 'disconnect', 'ping', 'wait_msg', 'keepalive', 'check_msg')
            print("MQTT (%s): %r" % (place_str[issue_place], conn_issue))

    def reconnect(self):
        """
        The function tries to resume the connection.

        Connection problems are captured and handled by `is_conn_issue()`
        """
        try:
            out = super().connect(False)
            self.conn_issue = None
            return out
        except (OSError, simple2.MQTTException) as e:
            self.conn_issue = (e, 4)
            if self.sock:
                self.sock.close()
                self.sock = None

    def resubscribe(self):
        """
        Function from previously registered subscriptions, sends them again to the server.

        :return:
        """
        for topic, qos in self.subs:
            self.subscribe(topic, qos, False)

    def add_msg_to_send(self, data):
        """
        By overwriting this method, you can control the amount of stored data in the queue.
        This is important because we do not have an infinite amount of memory in the devices.

        Currently, this method limits the queue length to MSG_QUEUE_MAX messages.

        The number of active messages is the sum of messages to be sent with messages awaiting confirmation.

        :param data:
        :return:
        """
        # Before we add data to the queue, it is necessary to release the memory first.
        # Otherwise, we may fall into an infinite loop due to a lack of available memory.
        messages_count = len(self.msg_to_send)
        messages_count += sum(map(len, self.msg_to_confirm.values()))

        while messages_count >= self.MSG_QUEUE_MAX:
            min_msg_to_confirm = min(map(lambda x: x[0] if x else 65535, self.msg_to_confirm.values()), default=0)
            if 0 < min_msg_to_confirm < 65535:
                key_to_check = None
                for k, v in self.msg_to_confirm.items():
                    if v and v[0] == min_msg_to_confirm:
                        del v[0]
                        key_to_check = k
                        break
                if key_to_check and key_to_check in self.msg_to_confirm and not self.msg_to_confirm[key_to_check]:
                    self.msg_to_confirm.pop(key_to_check)
            else:
                self.msg_to_send.pop(0)
            messages_count -= 1

        self.msg_to_send.append(data)

    def disconnect(self):
        """
        See documentation for `umqtt.simple2.MQTTClient.disconnect()`

        Connection problems are captured and handled by `is_conn_issue()`
        """
        try:
            return super().disconnect()
        except (OSError, simple2.MQTTException) as e:
            self.conn_issue = (e, 6)

    def ping(self):
        """
        See documentation for `umqtt.simple2.MQTTClient.ping()`

        Connection problems are captured and handled by `is_conn_issue()`
        """
        if not self.is_keepalive():
            return
        try:
            return super().ping()
        except (OSError, simple2.MQTTException) as e:
            self.conn_issue = (e, 7)

    def publish(self, topic, msg, retain=False, qos=0):
        """
        See documentation for `umqtt.simple2.MQTTClient.publish()`

        The function tries to send a message. If it fails, the message goes to the message queue for sending.

        The function does not support the `dup` parameter!

        When we have messages with the retain flag set, only one last message with that flag is sent!

        Connection problems are captured and handled by `is_conn_issue()`

        :return: None od PID for QoS==1 (only if the message is sent immediately, otherwise it returns None)
        """
        data = (topic, msg, retain, qos)
        if retain:
            # We delete all previous messages for this topic with the retain flag set to True.
            # Only the last message with this flag is relevant.
            self.msg_to_send[:] = [m for m in self.msg_to_send if not (topic == m[0] and retain == m[2])]
        try:
            out = super().publish(topic, msg, retain, qos, False)
            if qos == 1:
                # We postpone the message in case it is not delivered to the server.
                # We will delete it when we receive a receipt.
                self.msg_to_confirm.setdefault(data, []).append(out)
            return out
        except (OSError, simple2.MQTTException) as e:
            self.conn_issue = (e, 2)
            # If the message cannot be sent, we put it in the queue to try to resend it.
            if self.NO_QUEUE_DUPS:
                if data in self.msg_to_send:
                    return
            if self.KEEP_QOS0 and qos == 0:
                self.add_msg_to_send(data)
            elif qos == 1:
                self.add_msg_to_send(data)

    def subscribe(self, topic, qos=0, resubscribe=True):
        """
        See documentation for `umqtt.simple2.MQTTClient.subscribe()`

        The function tries to subscribe to the topic. If it fails,
        the topic subscription goes into the subscription queue.

        Connection problems are captured and handled by `is_conn_issue()`

        """
        data = (topic, qos)

        if self.RESUBSCRIBE and resubscribe:
            if topic not in dict(self.subs):
                self.subs.append(data)

        # We delete all previous subscriptions for the same topic from the queue.
        # The most important is the last subscription.
        self.sub_to_send[:] = [s for s in self.sub_to_send if topic != s[0]]
        try:
            out = super().subscribe(topic, qos)
            self.sub_to_confirm.setdefault(data, []).append(out)
            return out
        except (OSError, simple2.MQTTException) as e:
            self.conn_issue = (e, 3)
            if self.NO_QUEUE_DUPS:
                if data in self.sub_to_send:
                    return
            self.sub_to_send.append(data)

    def send_queue(self):
        """
        The function tries to send all messages and subscribe to all topics that are in the queue to send.

        :return: True if the queue's empty.
        :rtype: bool
        """
        msg_to_del = []
        for data in self.msg_to_send:
            topic, msg, retain, qos = data
            try:
                out = super().publish(topic, msg, retain, qos, False)
                if qos == 1:
                    # We postpone the message in case it is not delivered to the server.
                    # We will delete it when we receive a receipt.
                    self.msg_to_confirm.setdefault(data, []).append(out)
                msg_to_del.append(data)
            except (OSError, simple2.MQTTException) as e:
                self.conn_issue = (e, 5)
                return False
        self.msg_to_send[:] = [m for m in self.msg_to_send if m not in msg_to_del]
        del msg_to_del

        sub_to_del = []
        for data in self.sub_to_send:
            topic, qos = data
            try:
                out = super().subscribe(topic, qos)
                self.sub_to_confirm.setdefault(data, []).append(out)
                sub_to_del.append(data)
            except (OSError, simple2.MQTTException) as e:
                self.conn_issue = (e, 5)
                return False
        self.sub_to_send[:] = [s for s in self.sub_to_send if s not in sub_to_del]

        return True

    def is_conn_issue(self):
        """
        With this function we can check if there is any connection problem.

        It is best to use this function with the reconnect() method to resume the connection when it is broken.

        You can also check the result of methods such as this:
        `connect()`, `publish()`, `subscribe()`, `reconnect()`, `send_queue()`, `disconnect()`, `ping()`, `wait_msg()`,
        `check_msg()`, `is_keepalive()`.

        The value of the last error is stored in self.conn_issue.

        :return: Connection problem
        :rtype: bool
        """
        self.is_keepalive()

        if self.conn_issue:
            self.log()
        return bool(self.conn_issue)

    def wait_msg(self):
        """
        See documentation for `umqtt.simple2.MQTTClient.wait_msg()`

        Connection problems are captured and handled by `is_conn_issue()`
        """
        self.is_keepalive()
        try:
            return super().wait_msg()
        except (OSError, simple2.MQTTException) as e:
            self.conn_issue = (e, 8)

    def check_msg(self):
        """
        See documentation for `umqtt.simple2.MQTTClient.check_msg()`

        Connection problems are captured and handled by `is_conn_issue()`
        """
        self.is_keepalive()
        try:
            return super().check_msg()
        except (OSError, simple2.MQTTException) as e:
            self.conn_issue = (e, 10)
